example/bufferCopy/src/bufferCopy.cpp

```diff
@@ -55,6 +122,24 @@ struct PrintBufferKernel
     }
 };
 
+//#############################################################################
+//! Prints all elements of the buffer.
+struct PrintBufferKernel3D
+{
+    //-----------------------------------------------------------------------------
+    template<typename TAcc, typename TData, typename TExtent>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc, alpaka::accessor<TData, 3> const data, TExtent const& extents)
+        const -> void
+    {
+        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        auto const gridSize = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
+
+        for(size_t z = idx[0]; z < extents[0]; z += gridSize[0])
+            for(size_t y = idx[1]; y < extents[1]; y += gridSize[1])
+                for(size_t x = idx[2]; x < extents[2]; x += gridSize[2])
+                    printf("%zu,%zu,%zu:%u ", z, y, x, static_cast<uint32_t>(data(z, y, x)));
```

Comment:
Use `operator()(z, y, x)` to access the 3D buffer.

Reply:
I am against an interface where x, y, z are given seperate into the accessor exept there is also a vector/array (e.g. alpaka::vec) type interface available.

The reason is that if your app is written to support N dimensional buffers where you can choose N you can not easy write generic code.

Reply:
Yesterday it was late and I wrote from my smartphone, sp let me be more precise.
This code part is IMO not a good example because of the specialization to three-dimension you fastly assume that passing `x,y,z` separate is good. Additionally using three nested for loops will already limit your execution possibilities. This snipped is always sequential because for loops can not be parallelized if you do not use pragma based extensions.
Since this is outputting to the terminal the serialization is fine.
Think about you like to implement equally to this example an element-wise `copy`. You do not want to write a version for 1,2,3, ... dimensions so you would write.

pseudo code:
```C++
struct CopyElementWise
{
    //-----------------------------------------------------------------------------
    template<typename TAcc, typename TData, typename TExtent>
    ALPAKA_FN_ACC auto operator()(TAcc const& acc, alpaka::accessor<TData, 3> const data, TExtent const& extents)
        const -> void
    {
        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
        auto const gridSize = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
        auto globalIndex = calcGlobalIdx(gridSize,idx);
        destination[globalIndex] = data[globalIndex];
   }
};
```

This example kernel you would call as alpaka kernel. 
This example should demonstrate why I think an access method `operator[](vec)` should be prepared over `operator()(x,y,...)`

Reply:
> I am against an interface where x, y, z are given seperate into the accessor exept there is also a vector/array (e.g. alpaka::vec) type interface available.

That's ok, we should have an overload where we can supply Vec<N>. But I would also keep the overloads with explicit values for each dimension, because this is an easier API for normal people.

> The reason is that if your app is written to support N dimensional buffers where you can choose N you can not easy write generic code.

I think this is a very specific scenario and should not be the primary focus. But it does not cost much to support it.

> This snipped is always sequential because for loops can not be parallelized if you do not use pragma based extensions.

I think for loops inside an alpaka kernel do not need to parallelize, because we already parallelized the outer loops (blocks, threads, elements).

>         auto globalIndex = calcGlobalIdx(gridSize,idx);
>       destination[globalIndex] = data[globalIndex];

What does globalIndex do here? Is it a type holding 3 integers? Because we should avoid mapping a 3D index into a 1D index. This is done by the accessor not by the user code.

Reply:
> What does globalIndex do here? Is it a type holding 3 integers? Because we should avoid mapping a 3D index into a 1D index. This is done by the accessor not by the user code.

`globalIndex` is in my pseudo code an N-dimensional object e.g alpaka:Vec.